packages/iov-stream/src/defaultvalueproducer.spec.ts

```diff
@@ -0,0 +1,95 @@
+import { Stream } from "xstream";
+
+import { DefaultValueProducer } from "./defaultvalueproducer";
+
+function oneTickLater(): Promise<void> {
+  return new Promise(resolve => setTimeout(resolve));
+}
+
+describe("DefaultValueProducer", () => {
+  it("can be constructed", () => {
+    const producer = new DefaultValueProducer(1);
+    expect(producer.value).toEqual(1);
+  });
+
+  it("can be used as a stream backend", done => {
+    const producer = new DefaultValueProducer(42);
+    const stream = Stream.createWithMemory(producer);
+    stream.addListener({
+      next: value => {
+        expect(value).toEqual(42);
+        done();
+      },
+      error: fail,
+      complete: fail,
+    });
+  });
+
+  it("can send updates", done => {
+    const producer = new DefaultValueProducer(42);
+    const stream = Stream.createWithMemory(producer);
+
+    // tslint:disable-next-line:readonly-array
+    const events: number[] = [];
+    stream.addListener({
+      next: value => {
+        events.push(value);
+
+        if (events.length === 4) {
+          expect(events).toEqual([42, 43, 44, 45]);
+          done();
+        }
+      },
+      error: fail,
+      complete: fail,
+    });
+
+    producer.update(43);
+    producer.update(44);
+    producer.update(45);
+  });
+
+  it("calls callbacks", async () => {
+    // tslint:disable-next-line:readonly-array
+    const producerActive: boolean[] = [];
+
+    const producer = new DefaultValueProducer(42, {
+      onStarted: () => producerActive.push(true),
+      onStop: () => producerActive.push(false),
+    });
+    const stream = Stream.createWithMemory(producer);
+
+    expect(producerActive).toEqual([]);
+
+    const subscription1 = stream.subscribe({});
+    expect(producerActive).toEqual([true]);
+
+    const subscription2 = stream.subscribe({});
+    expect(producerActive).toEqual([true]);
```

Comment:
This is because the producer is started as soon as 1 or more subscriptions exist. At that point you can add as many subscriptions to the producer as you want. When the amount of subscriptions drops to 0, the producer stops.

The `producerActive` array is a list of _one or more are listening_ and _no one is listening_ values.

Any idea how to improve readability? I could create an enum:

```
enum SomeoneListening {
  Yes,
  No
}
```

instead of the boolean. Do you think that helps?

Reply:
It looks good to me as is.

I just had to scroll up to see the push onStart and onStop and it makes sense. Since that is obvious effect, I'd leave it as is.